In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot

In [3]:
dbt=pd.read_csv('Diabetes (1).csv')
dbt.head()

Number of times pregnant   Plasma glucose concentration  \
0                          6                            148   
1                          1                             85   
2                          8                            183   
3                          1                             89   
4                          0                            137   

    Diastolic blood pressure   Triceps skin fold thickness  \
0                         72                            35   
1                         66                            29   
2                         64                             0   
3                         66                            23   
4                         40                            35   

    2-Hour serum insulin   Body mass index   Diabetes pedigree function  \
0                      0              33.6                        0.627   
1                      0              26.6                        0.351   
2                      0              23.3                        0.672   
3                     94              28.1                        0.167   
4                    168              43.1                        2.288   

    Age (years)  Class variable  
0            50             YES  
1            31              NO  
2            32             YES  
3            21              NO  
4            33             YES

In [4]:
dbt.describe()

Number of times pregnant   Plasma glucose concentration  \
count                 768.000000                     768.000000   
mean                    3.845052                     120.894531   
std                     3.369578                      31.972618   
min                     0.000000                       0.000000   
25%                     1.000000                      99.000000   
50%                     3.000000                     117.000000   
75%                     6.000000                     140.250000   
max                    17.000000                     199.000000   

        Diastolic blood pressure   Triceps skin fold thickness  \
count                 768.000000                    768.000000   
mean                   69.105469                     20.536458   
std                    19.355807                     15.952218   
min                     0.000000                      0.000000   
25%                    62.000000                      0.000000   
50%                    72.000000                     23.000000   
75%                    80.000000                     32.000000   
max                   122.000000                     99.000000   

        2-Hour serum insulin   Body mass index   Diabetes pedigree function  \
count             768.000000        768.000000                   768.000000   
mean               79.799479         31.992578                     0.471876   
std               115.244002          7.884160                     0.331329   
min                 0.000000          0.000000                     0.078000   
25%                 0.000000         27.300000                     0.243750   
50%                30.500000         32.000000                     0.372500   
75%               127.250000         36.600000                     0.626250   
max               846.000000         67.100000                     2.420000   

        Age (years)  
count    768.000000  
mean      33.240885  
std       11.760232  
min       21.000000  
25%       24.000000  
50%       29.000000  
75%       41.000000  
max       81.000000

In [5]:
dbt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0    Number of times pregnant      768 non-null    int64  
 1    Plasma glucose concentration  768 non-null    int64  
 2    Diastolic blood pressure      768 non-null    int64  
 3    Triceps skin fold thickness   768 non-null    int64  
 4    2-Hour serum insulin          768 non-null    int64  
 5    Body mass index               768 non-null    float64
 6    Diabetes pedigree function    768 non-null    float64
 7    Age (years)                   768 non-null    int64  
 8    Class variable                768 non-null    object 
dtypes: float64(2), int64(6), object(1)
memory usage: 54.1+ KB


In [11]:
# Normalization function 
def norm_func(i):
    x = (i-i.min())	/ (i.max()-i.min())
    return (x)

In [12]:
# Normalized data frame (considering the numerical part of data)
dbt_n = norm_func(dbt.iloc[:,:8])
dbt_n.describe()

Number of times pregnant   Plasma glucose concentration  \
count                 768.000000                     768.000000   
mean                    0.226180                       0.607510   
std                     0.198210                       0.160666   
min                     0.000000                       0.000000   
25%                     0.058824                       0.497487   
50%                     0.176471                       0.587940   
75%                     0.352941                       0.704774   
max                     1.000000                       1.000000   

        Diastolic blood pressure   Triceps skin fold thickness  \
count                 768.000000                    768.000000   
mean                    0.566438                      0.207439   
std                     0.158654                      0.161134   
min                     0.000000                      0.000000   
25%                     0.508197                      0.000000   
50%                     0.590164                      0.232323   
75%                     0.655738                      0.323232   
max                     1.000000                      1.000000   

        2-Hour serum insulin   Body mass index   Diabetes pedigree function  \
count             768.000000        768.000000                   768.000000   
mean                0.094326          0.476790                     0.168179   
std                 0.136222          0.117499                     0.141473   
min                 0.000000          0.000000                     0.000000   
25%                 0.000000          0.406855                     0.070773   
50%                 0.036052          0.476900                     0.125747   
75%                 0.150414          0.545455                     0.234095   
max                 1.000000          1.000000                     1.000000   

        Age (years)  
count    768.000000  
mean       0.204015  
std        0.196004  
min        0.000000  
25%        0.050000  
50%        0.133333  
75%        0.333333  
max        1.000000

In [17]:
##Splitting the data into featuers and labels
X = features = dbt_n
Y = labels = dbt.iloc[:,8]

In [15]:
features

Number of times pregnant   Plasma glucose concentration  \
0                     0.352941                       0.743719   
1                     0.058824                       0.427136   
2                     0.470588                       0.919598   
3                     0.058824                       0.447236   
4                     0.000000                       0.688442   
..                         ...                            ...   
763                   0.588235                       0.507538   
764                   0.117647                       0.613065   
765                   0.294118                       0.608040   
766                   0.058824                       0.633166   
767                   0.058824                       0.467337   

      Diastolic blood pressure   Triceps skin fold thickness  \
0                     0.590164                      0.353535   
1                     0.540984                      0.292929   
2                     0.524590                      0.000000   
3                     0.540984                      0.232323   
4                     0.327869                      0.353535   
..                         ...                           ...   
763                   0.622951                      0.484848   
764                   0.573770                      0.272727   
765                   0.590164                      0.232323   
766                   0.491803                      0.000000   
767                   0.573770                      0.313131   

      2-Hour serum insulin   Body mass index   Diabetes pedigree function  \
0                 0.000000          0.500745                     0.234415   
1                 0.000000          0.396423                     0.116567   
2                 0.000000          0.347243                     0.253629   
3                 0.111111          0.418778                     0.038002   
4                 0.198582          0.642325                     0.943638   
..                     ...               ...                          ...   
763               0.212766          0.490313                     0.039710   
764               0.000000          0.548435                     0.111870   
765               0.132388          0.390462                     0.071307   
766               0.000000          0.448584                     0.115713   
767               0.000000          0.453055                     0.101196   

      Age (years)  
0        0.483333  
1        0.166667  
2        0.183333  
3        0.000000  
4        0.200000  
..            ...  
763      0.700000  
764      0.100000  
765      0.150000  
766      0.433333  
767      0.033333  

[768 rows x 8 columns]

In [16]:
labels

0      YES
1       NO
2      YES
3       NO
4      YES
      ... 
763     NO
764     NO
765     NO
766    YES
767     NO
Name:  Class variable, Length: 768, dtype: object

In [21]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)

In [22]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(n_jobs=2, n_estimators=15, criterion="entropy")

In [24]:
rf.fit(X_train, Y_train) # Fitting RandomForestClassifier model from sklearn.ensemble  
pred = rf.predict(X_test)
pred

array(['NO', 'NO', 'NO', 'NO', 'NO', 'NO', 'NO', 'NO', 'NO', 'NO', 'NO',
       'NO', 'NO', 'YES', 'NO', 'YES', 'NO', 'YES', 'YES', 'NO', 'NO',
       'YES', 'NO', 'NO', 'NO', 'YES', 'YES', 'NO', 'NO', 'NO', 'YES',
       'NO', 'NO', 'YES', 'YES', 'NO', 'YES', 'YES', 'NO', 'YES', 'NO',
       'NO', 'YES', 'YES', 'YES', 'NO', 'NO', 'NO', 'NO', 'YES', 'YES',
       'YES', 'NO', 'NO', 'NO', 'YES', 'NO', 'NO', 'YES', 'NO', 'YES',
       'NO', 'NO', 'NO', 'YES', 'NO', 'NO', 'NO', 'NO', 'NO', 'NO', 'NO',
       'NO', 'YES', 'YES', 'YES', 'NO', 'NO', 'NO', 'YES', 'NO', 'NO',
       'NO', 'NO', 'NO', 'YES', 'NO', 'YES', 'NO', 'NO', 'YES', 'NO',
       'NO', 'NO', 'YES', 'NO', 'YES', 'NO', 'NO', 'NO', 'NO', 'YES',
       'YES', 'NO', 'YES', 'YES', 'YES', 'YES', 'YES', 'NO', 'NO', 'NO',
       'YES', 'NO', 'NO', 'NO', 'YES', 'YES', 'NO', 'NO', 'NO', 'NO',
       'YES', 'YES', 'NO', 'YES', 'NO', 'NO', 'NO', 'NO', 'YES', 'NO',
       'YES', 'NO', 'YES', 'NO', 'NO', 'YES', 'NO', 'NO', 'YES', 'NO',


In [26]:
from sklearn.metrics import accuracy_score

In [27]:
pd.crosstab(Y_test, pred, rownames=['Actual'], colnames= ['Predictions']) 
print(accuracy_score(Y_test, pred))

0.7056277056277056


Model can predict the test data with 70.5% accuracy

In [28]:
# test accuracy
test_acc2 = np.mean(rf.predict(X_test)==Y_test)
test_acc2

0.7056277056277056

In [29]:
# train accuracy 
train_acc2 = np.mean(rf.predict(X_train)==Y_train)
train_acc2

0.9962756052141527

Model can predict the training data with 99.6% accuracy

In [31]:
rf.estimators_

[DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=457402290),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=829161727),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=853599328),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=496021882),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=1702567987),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=765812478),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=640459653),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
                        random_state=1558647955),
 DecisionTreeClassifier(criterion='entropy', max_features='auto',
            

In [32]:
rf.classes_

array(['NO', 'YES'], dtype=object)

In [33]:
rf.n_features_

8

In [34]:
rf.n_classes_

2

In [36]:
rf.n_outputs_

1

In [40]:
tree = rf.estimators_[5]
tree

DecisionTreeClassifier(criterion='entropy', max_features='auto',
                       random_state=765812478)